src/Sarif.Converters/ToolFileConverterBase.cs

```diff
@@ -75,5 +75,54 @@ internal static string GetLogicalLocationName(string parentKey, string fullyQual
 
             return logicalName;
         }
+
+        protected static Run PersistResults(IResultLogWriter output, IList<Result> results, string toolName)
+        {
+            var run = new Run()
+            {
+                Tool = new Tool { Name = toolName },
+                ColumnKind = ColumnKind.Utf16CodeUnits
+            };
+
+            return PersistResults(output, results, run);
+        }
+
+        protected static Run PersistResults(IResultLogWriter output, IList<Result> results, Run run)
+        {
+            output.Initialize(run);
+
+            if (run.Invocations.Any())
+            {
+                // TODO: add WriteInvocations to IResultLogWriter
+                // https://github.com/Microsoft/sarif-sdk/issues/1190
+                (output as ResultLogJsonWriter).WriteInvocations(run.Invocations);
+            }
+
+            run.Results = results;
+            var visitor = new AddFileReferencesVisitor();
+            visitor.VisitRun(run);
+
+            if (run.Files.Any())
+            {
+                output.WriteFiles(run.Files);
+            }
+
+            if (run.LogicalLocations.Any())
+            {
+                output.WriteLogicalLocations(run.LogicalLocations);
+            }
+
+            output.OpenResults();
+            output.WriteResults(run.Results);
+            output.CloseResults();
+
+            // TODO: should we move these to be emitted at the beginning of the file? What order fo we prefer generaally
+            if (run.Resources?.Rules != null)
```

Comment:
>if [](start = 12, length = 2)

A tool that emits results as it produces them would naturally put `files` and `logicalLocations` at the end, because doesn't know until the end the set of locations it will encounter. Such a tool would also put `resources` at the end if it chose to emit only those rules that actually triggered and those messages it actually emitted; otherwise it could put `resources` anywhere.

Also, a user probably wants to get to the results ASAP, without paging through dozens thousands of files before getting to the good stuff.

So IMO all of `files`, `logicalLocations`, and `resources` should go after `results`.

Reply:
Just to be clear, we're only discussing the converters scenario, not tools that emits results as they produce them. Now that we've DRYed out this code, we have the possibility of making all converter behaviors consistent. I'll put everything afterwards as you say. 

---
In reply to: [244520670](https://github.com/Microsoft/sarif-sdk/pull/1191#discussion_r244520670) [](ancestors = 244520670)

Reply:
but btw one consequence of this is that our next converter update will produce wildly divergent content that will be impossible to diff effectively manually. Even if you are Laurence Golding, esq.

---
In reply to: [244545376](https://github.com/Microsoft/sarif-sdk/pull/1191#discussion_r244545376) [](ancestors = 244545376,244520670)